Add ViTMatte

NielsRogge opened the pull request:
# What does this PR do?

This PR supersedes #25051 and #25524. 

To do:

- [x] simplify image processor
- [x] fix image processor tests => @amyeroberts should I overwrite the following tests, since the image processor also requires trimaps to be passed?

```
FAILED tests/models/vitmatte/test_image_processing_vitmatte.py::VitMatteImageProcessingTest::test_call_numpy - TypeError: preprocess() missing 1 required positional argument: 'trimaps'
FAILED tests/models/vitmatte/test_image_processing_vitmatte.py::VitMatteImageProcessingTest::test_call_numpy_4_channels - TypeError: preprocess() missing 1 required positional argument: 'trimaps'
FAILED tests/models/vitmatte/test_image_processing_vitmatte.py::VitMatteImageProcessingTest::test_call_pil - TypeError: preprocess() missing 1 required positional argument: 'trimaps'
FAILED tests/models/vitmatte/test_image_processing_vitmatte.py::VitMatteImageProcessingTest::test_call_pytorch - TypeError: preprocess() missing 1 required positional argument: 'trimaps'
FAILED tests/models/vitmatte/test_image_processing_vitmatte.py::VitMatteImageProcessingTest::test_image_processor_from_dict_with_kwargs - AssertionError: 42 != {'height': 42, 'width': 42}
```

NielsRogge committed:
First draft

NielsRogge committed:
Simplify image processor

NielsRogge committed:
Fix rebase

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,200 @@
+        self.image_std = image_std if image_std is not None else IMAGENET_STANDARD_STD
+        self.size_divisibility = size_divisibility
+
+    def pad(self, images: np.ndarray, size_divisibility: int = 32):
```
Missing doc string

amyeroberts replied:
All image_processor methods should accept data_format and input_data_format as arguments

amyeroberts replied:
This'll need to be named pad_image (because old feature extractor methods used `pad` for taking batches of images and returning BatchFeatures)

amyeroberts commented on tests/models/vitmatte/test_image_processing_vitmatte.py:
```diff
@@ -0,0 +1,106 @@
+# coding=utf-8
+# Copyright 2021 HuggingFace Inc.
```
```suggestion
# Copyright 2023 HuggingFace Inc.
```

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,200 @@
+
+        Args:
+            images (`ImageInput`):
+                Image to preprocess.
```
Should add that input is expected to have pixel values of 0-255

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,200 @@
+
+        if do_rescale and rescale_factor is None:
+            raise ValueError("Rescale factor must be specified if do_rescale is True.")
+
```
Check on whether the input should be rescaled here

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,200 @@
+                "torch.Tensor, tf.Tensor or jax.ndarray."
+            )
+
+        if do_rescale and rescale_factor is None:
```
Equivalent checks for `do_normalize` and `do_pad` should be added here

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,200 @@
+            images = self.pad(np.array(images), size_divisibility=size_divisibility)
+
+        images = [
+            to_channel_dimension_format(image=image, channel_dim=data_format, input_channel_dim=ChannelDimension.LAST)
```
```suggestion
            to_channel_dimension_format(image=image, channel_dim=data_format, input_channel_dim=input_data_format)
```

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,200 @@
+        self.size_divisibility = size_divisibility
+
+    def pad(self, images: np.ndarray, size_divisibility: int = 32):
+        batch_size, height, width, num_channels = images.shape
```
All the transforms should take a single image (or concatenation of image and trimap) and work with images in channels first and channels last format

NielsRogge replied:
Ok, however isn't it more efficient to pad a single batch rather than doing it one by one?

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+    """
+
+    loss: Optional[torch.FloatTensor] = None
+    alphas: torch.FloatTensor = None
```
I'd just call these logits to keep consistency with other model outputs 

```suggestion
    logits: torch.FloatTensor = None
```

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+        padding=1,
+    ):
+        super().__init__()
+        self.conv = nn.Conv2d(in_channels, out_channels, 3, stride, padding, bias=False)
```
3 should be a kwarg to make it clear what it's setting 

```suggestion
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size3, stride=stride, padding=padding, bias=False)
```

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+    ):
+        super().__init__()
+        self.conv = nn.Conv2d(in_channels, out_channels, 3, stride, padding, bias=False)
+        self.bn = nn.BatchNorm2d(out_channels)
```
Let's use a clearer name than `bn`

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+        self,
+        in_channels,
+        out_channels,
+        stride=2,
+        padding=1,
```
We don't need these args as they're always the same value 

```suggestion
```

NielsRogge replied:
Actualy the `VitMatteBasicConv3x3` is used in 2 different places, with a different stride.

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+            raise ValueError("The length of fusion_out should be equal to the length of convstream_out + 1.")
+
+        self.config = config
+        self.convstream = VitMatteConvStream(
+            in_channels=config.backbone_config.num_channels, out_channels=config.convstream_out
+        )
```
This layer should just take the config and use that to instantiate itself

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+    Simple ConvStream containing a series of basic conv3x3 layers to extract detail features.
+    """
+
+    def __init__(self, in_channels=4, out_channels=[48, 96, 192]):
```
We shouldn't use mutables as default values

amyeroberts replied:
Should take config as its first argument

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+    Simple fusion block to fuse features from ConvStream and Plain Vision Transformer.
+    """
+
+    def __init__(self, in_channels, out_channels):
```
All layers should take the config its first argument

NielsRogge replied:
This fusion block has different in and out channels depending on the layer idx, so it cannot define it based on the config, unless we add a `layer_idx` argument

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+    Basic convolution layers including: Conv3x3, BatchNorm2d, ReLU layers.
+    """
+
+    def __init__(
```
Should take config as its first argument

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+    Simple Matting Head, containing only conv3x3 and conv1x1 layers.
+    """
+
+    def __init__(self, in_channels=32, mid_channels=16):
```
Should take `config`

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+    def __init__(self, in_channels=32, mid_channels=16):
+        super().__init__()
+        self.matting_convs = nn.Sequential(
+            nn.Conv2d(in_channels, mid_channels, 3, 1, 1),
```
The numerical values - 3, 1, 1 should be passed in as kwargs so it's clear what they're setting

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+        super().__init__()
+        self.convs = nn.ModuleList()
+
+        self.conv_chans = out_channels.copy()
+        self.conv_chans.insert(0, in_channels)
```
This is super dirty - please don't do this! 

We can have `out_channels` as `None` which default to the desired list

```python
out_channels = [in_channels, 48, 96, 192] if out_channels is None else out_channels
```

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,282 @@
+        super().__init__()
+        self.conv = VitMatteBasicConv3x3(in_channels, out_channels, stride=1, padding=1)
+
+    def forward(self, features, D):
```
Clearer var name than `D` should go here

amyeroberts reviewed (commented):
Thanks for adding this model! 

I've done an initial pass. Mostly looks good, but there's a few places which need to be updated before it's transformers ready. Once those comments have been resolved and the tests are all passing ping me again for another review :) 

cc @rafaelpadilla

NielsRogge commented on docs/source/en/model_doc/vitmatte.md:
```diff
@@ -0,0 +1,44 @@
+## Overview
+
+The VitMatte model was proposed in [Boosting Image Matting with Pretrained Plain Vision Transformers](https://arxiv.org/abs/2305.15272) by Jingfeng Yao, Xinggang Wang, Shusheng Yang, Baoyuan Wang.
+VitMatte leverages plain [Vision Transformers](vit) for the task of image matting, which is the process of accurately estimating the foreground object in images and videos.
```
The sentence above is the one-line summary:

> VitMatte leverages plain [Vision Transformers](vit) for the task of image matting, which is the process of accurately estimating the foreground object in images and videos.

NielsRogge commented on src/transformers/models/vitmatte/configuration_vitmatte.py:
```diff
@@ -0,0 +1,102 @@
+
+    def __init__(
+        self,
+        backbone_config=None,
```
Could you clarify? This is not done yet for any of our configs, e.g. https://github.com/huggingface/transformers/blob/df04959e5542d41b269f96305d82c62287350cee/src/transformers/models/upernet/configuration_upernet.py#L76

amyeroberts replied:
Type hints should be added for all our public classes (it's good practice to be adding them generally). It's true we don't have then for existing configs - but we should start adding them.

NielsRogge committed:
Address comments

NielsRogge committed:
Address more comments

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25843). All of your documentation changes will be reflected on that endpoint.

NielsRogge committed:
Address more comments

NielsRogge committed:
Address more comments

NielsRogge committed:
Address more comments

NielsRogge committed:
Improve pad_image

NielsRogge committed:
Add tests

NielsRogge committed:
Update integration test

NielsRogge committed:
Fix image processor tests

NielsRogge committed:
Fix model tests

NielsRogge committed:
Convert checkpoints

NielsRogge committed:
Fix doc tests

NielsRogge committed:
Remove file

NielsRogge commented:
@amyeroberts thanks for your review, I've addressed all comments. However regarding renaming the outputs from `alphas` to `logits` => are we sure we want to do this, given that image matting models output the alpha values, not really logits?

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,266 @@
+        if input_data_format is None:
+            input_data_format = infer_channel_dimension_format(image)
+
+        height, width = image.shape[:2] if input_data_format == ChannelDimension.LAST else image.shape[1:]
```
```suggestion
        height, width = get_image_size(image, input_data_format)
```

amyeroberts commented on src/transformers/models/vitmatte/image_processing_vitmatte.py:
```diff
@@ -0,0 +1,266 @@
+        size_divisibility: int = 32,
+        data_format: Optional[Union[str, ChannelDimension]] = None,
+        input_data_format: Optional[Union[str, ChannelDimension]] = None,
+    ):
```
```suggestion
    ) -> np.ndarray:
```

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,341 @@
+            detailed_feature_map_name = "detailed_feature_map_" + str(len(self.fusion_blocks) - i - 1)
+            features = self.fusion_blocks[i](features, detail_features[detailed_feature_map_name])
+
+        phas = torch.sigmoid(self.matting_head(features))
+
+        return phas
```
```suggestion
        alphas = torch.sigmoid(self.matting_head(features))

        return alphas
```

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,341 @@
+    Basic convolution layers including: Conv3x3, BatchNorm2d, ReLU layers.
+    """
+
+    def __init__(self, in_channels, out_channels, stride=2, padding=1):
```
Should still take the config as its first argument here

NielsRogge replied:
See my comment below for why it's not needed

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,341 @@
+            padding=padding,
+            bias=False,
+        )
+        self.batch_norm = nn.BatchNorm2d(out_channels)
```
Should use the config batch norm eps here

NielsRogge replied:
There's no batch norm eps config attribute. Is this necessary to be added?

amyeroberts replied:
Yes, it's the same for layer norm eps, it should be added.

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,341 @@
+            bias=False,
+        )
+        self.batch_norm = nn.BatchNorm2d(out_channels)
+        self.relu = nn.ReLU(True)
```
Why set to inplace here?

amyeroberts reviewed (commented):
Thanks for iterating! 

Just a few small changes and I think we're good to merge. It will be great to have this model added! 

Re alphas vs logits in the model output. Happy for it to be alphas as we apply a sigmoid on the output so it's [0, 1]

amyeroberts commented:
cc @rafaelpadilla for reference

NielsRogge committed:
Apply suggestions

NielsRogge committed:
Address comments

NielsRogge committed:
Fix typing hint

NielsRogge committed:
Add batch_norm_eps

NielsRogge commented:
Addressed all comments. Feel free to merge if approved

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,342 @@
+    Args:
+        pixel_values (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Pixel values. Padding will be ignored by default should you provide it. Pixel values can be obtained using
+            [`AutoImageProcessor`]. See [`SegformerImageProcessor.__call__`] for details.
```
```suggestion
            [`AutoImageProcessor`]. See [`VitMatteImageProcessor.__call__`] for details.
```

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -0,0 +1,342 @@
+    Simple Matting Head, containing only conv3x3 and conv1x1 layers.
+    """
+
+    def __init__(self, config, mid_channels=16):
```
We can remove `mid_channels` as it's always the same

amyeroberts reviewed (approved):
Thanks for adding this model! 

Just a tiny tiny nit. @rafaelpadilla could you handle merging this PR once resolved?

NielsRogge committed:
Address comments

NielsRogge committed:
Fix style

NielsRogge commented:
Failing tests are flaky/unrelated

rafaelpadilla reviewed (commented):
Looks great to me! :) 
Excellent work, @NielsRogge

rafaelpadilla commented:
@NielsRogge 
It seems to me that every comment was addressed and now all tests are passing.
I believe now it is ready to merge. May I?

NielsRogge commented:
Yes
